numpy/linalg/__init__.py

```diff
@@ -50,6 +50,28 @@
 
 from .linalg import *
 
+import numpy.distutils.system_info as sinfo
+if len(sinfo.get_info('openblas')) > 2:
+    # openblas >= 0.3.4 provides version information;
+    # numpy/distutils/system_info has already initialized
+    # at this stage, so just update the dictionary for
+    # openblas info
+    from numpy.linalg import openblas_config
+    openblas_config_str = openblas_config._openblas_info()
+    openblas_config_list = openblas_config_str.split()
+    # slightly obscure API for system_info objects;
+    # they weren't really intended for public modification
+    # as the docs clearly indicate
+    info = sinfo.system_info.saved_results['openblas_info']
+    new_info = {}
+    if openblas_config_list[0] == b"OpenBLAS":
+        # version string will be present
+        new_info['version'] = openblas_config_list[1]
+    else:
+        # older OpenBLAS config API
+        new_info['version'] = None
+    info.update(new_info)
+
```

Comment:
Could you replace this with a ctypes-based check in [`openblas_info.calc_info`](https://github.com/numpy/numpy/blob/v1.16.0rc2/numpy/distutils/system_info.py#L1756)? You would need to find the exact shared object name sucessfully found in `self.check_libs`, but then the rest is easy (assuming ctypes exists, which I think is true for any python version we care about)
```
dll = ctypes.CDLL(lib_name)
openblas_get_config = dll.openblas_get_config
openblas_get_config.restype = ctypes.c_char_p
openblas_config_str = openblas_get_config()
```

Reply:
I'm not sure I understand how this helps. You're certainly right that it would be nice to intervene in the `system_info.py` module, and even better so if the initial detection of `OpenBLAS` is immediately flagged / logged with the version during the NumPy build.

But the problem is compiling the C module into the NumPy namespace inline in the Python code in `system_info.py`, or somewhere upstream in the build process, and I think this Python code is used extremely early in the NumPy build control flow. It is harder to go much farther upstream in the build process than the detection of the libraries though, which I think is why it is just so much easier to compile from a NumPy submodule "downstream" of distutils in the build process.

Reply:
In other words, in your suggestion, how do I generate the shared object file upstream in the workflow? Maybe just the `try_link` or `try_compile` stuff still is a possibility if I can hook into the return values.

Reply:
Ah, cool, you're referring to openblas itself instead of a compiled file associated with the new C extension. I should definitely try that since compiling the new C extension is the major bottleneck.

Reply:
Seems to work well in my hands for Mac / Linux, but for Windows I'm having to think a bit more because we link with the static library there. Maybe I can cheat and use the dynamic libraries that come in the bundle just for the versioning purpose.

Reply:
Strange, it seems the windows wheels for 1.15.4 dynamically link to an OpenBLAS dll. Are the wheels different from the dev builds?

Reply:
https://github.com/MacPython/numpy-wheels/blob/master/.appveyor.yml#L89

Reply:
That's the static lib used for wheels too -- we even physically remove the dynamic libs for good measure

Reply:
I think there's a good reason for that too -- I think it is related to Windows + mingw + openblas combination and static being required for some reason. Not sure if that is still true, but pretty sure we still do it, and changing that in the openblas wheel generation workflow could cause quite a bit of disruption.

There is a `dll.a` file available as well on Windows, but so far `ctypes` doesn't like it either, and I think those may just be helper files for when the true dynamic lib is available.

Reply:
So, I think it is back to the original issue -- fully portable detection of static / dynamic openblas lib version contained to distutils is hard. Using C extension is more portable, but hard to confine to distutils.

Reply:
Well, when I download the [wheel for 1.16rc2](https://files.pythonhosted.org/packages/8a/51/44662dbc1963cd1e542459aa75e212cb9f1783b84b88192d33bf0d00b78c/numpy-1.16.0rc2-cp27-cp27m-win_amd64.whl) and unzip it, I find `numpy/.libs/libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll`. 

Using `depends.exe` I see that `numpy/core/_multiarray_umath.pyd` depends on that `dll`.

Perhaps the result of `self.check_libs` gives the import library rather than the dll, since MSVC links to a `*.lib` file not a `*.dll` file. At runtime there is an algorithm to determine where to fine the matching `*.dll`. listed inside the import library

Reply:
The real question for me is whether the Windows openblas dynamic library is available before NumPy even starts compiling. So from our Azure openblas zip URL I download from here and unzip: https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-v0.3.3-186-g701ea883-win_amd64-gcc_7_1_0.zip

I don't think anything in there was usable with `cytpes` for dynamic library interaction in my hands. If there's an intermediate command needed to generate the requisite ctypes-tractable file first then I'd be glad to know it.

@matthew-brett Do you know offhand if these Windows openblas zip files are supposed to have dynamic libraries ready for i.e., ctypes access?

Reply:
I certainly saw way more files for the MacOS version of the ZIP--in fact the code I wrote to parse for version number goes through the various files available in the zip and ignores the ones that don't work with ctypes--only a few of the files for Mac worked, and none of the Windows ones.

Reply:
I do see `libopenblas_v0.3.3-186-g701ea883-gcc_7_1_0.dll` actually -- it is hiding in the `bin` path instead of `lib`, so I can play with that maybe!

Reply:
The [azure build script](https://github.com/numpy/numpy/blob/v1.16.0rc2/azure-pipelines.yml#L160) only copies the `*.a` file from the zip. I wonder where the `/libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll` file that is in the wheel comes from?

Reply:
@mattip Does that work for you for the DLL in the openblas zip file mentioned above?

Reply:
on a Windows machine

Reply:
@tylerjereddy: yes

Reply:
even `depends.exe` used in combination with the absolute file path or in the same directory as the OpenBLAS DLL produces "The system cannot find the file specified (2)." after Dependency Walker loads.

I need a way to get reliable missing DLL / dependencies on the openblas DLL I download, irrespective of my current path / other system configs, assuming this is the problem.